### PR TITLE
Remote artifact repository bi-directional sync

### PR DESCRIPTION
**Pull request by Equanox**

### CODE REVIEW

File: bobtask/artifact.go [hunk 1]
```diff
@@ -73,7 +73,11 @@ func (t *Task) ArtifactPack(artifactName hash.In) (err error) {
 		exports = append(exports, filepath.Join(t.dir, path.String()))
 	}
 
+	// TODO: reestablish the local store.
 	artifact, err := t.local.NewArtifact(context.TODO(), artifactName.String())
+
+	// TODO: Currently the remote store ignores the username.
+	// artifact, err := t.remote.NewArtifact(context.TODO(), artifactName.String())
```
Comment (rdnt) [inline bobtask/artifact.go:80]: I suppose this can be removed? (We sync at the end anyway)

File: bob/bobfile/bobfile.go [hunk 2]
```diff
@@ -182,6 +216,24 @@ func initializeDependencies(dir string, task bobtask.Task, bobfile *Bobfile) []n
...
+	parts := strings.Split(strings.TrimLeft(endpoint.Path, sep), sep)
+
+	username := parts[0]
+	project := strings.Join(parts[1:], sep)
+
+	s = remotestore.New(
+		username,
+		project,
+		remotestore.WithClient(
+			storeclient.New("http://"+endpoint.Host),
```
Comment (Equanox) [inline bob/bobfile/bobfile.go:231]: That should be https by default. We should add a explicit option to allow insecure.

File: bob/build.go [hunk 3]
```diff
@@ -39,8 +43,61 @@ func (b *B) Build(ctx context.Context, taskName string) (err error) {
...
+		}
+	}
+
 	err = playbook.Build(ctx)
 	errz.Fatal(err)
 
-	return err
+	// sync artifacts from current build with remote store
+	if remotestore != nil {
+		artifactIds := []hash.In{}
```
Comment (Equanox) [inline bob/build.go:75]: Shall we put this block into a function?

File: bob/build.go [hunk 4]
```diff
@@ -39,8 +43,61 @@ func (b *B) Build(ctx context.Context, taskName string) (err error) {
 	err = playbook.PreComputeInputHashes()
 	errz.Fatal(err)
 
+	remotestore := ag.Remotestore()
+	if remotestore != nil {
+		artifactIds := []hash.In{}
```
Comment (Equanox) [inline bob/build.go:48]: Shall we put this block into a function?

File: bobtask/artifact.go (hunk 1 above)
Comment (Equanox) [reply bobtask/artifact.go:80]: Yes pleas remove.

File: bobtask/artifact.go [hunk 5]
```diff
@@ -73,7 +73,11 @@ func (t *Task) ArtifactPack(artifactName hash.In) (err error) {
 		exports = append(exports, filepath.Join(t.dir, path.String()))
 	}
 
+	// TODO: reestablish the local store.
```
Comment (Equanox) [inline bobtask/artifact.go:76]: remove

File: pkg/store/remotestore/remotestore.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+			s.project,
+			artifactID,
+			reader,
+		)
+		if err != nil {
+			//errz.Log(err)
+			s.err = err
+		}
+
+		//_ = writer.CloseWithError(err)
```
Comment (Equanox) [inline pkg/store/remotestore/remotestore.go:67]: Remove

File: pkg/store/remotestore/remotestore.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+	go func() {
+		defer s.wg.Done()
+		err := s.client.UploadArtifact(
+			ctx,
+			s.project,
+			artifactID,
+			reader,
+		)
+		if err != nil {
+			//errz.Log(err)
```
Comment (Equanox) [inline pkg/store/remotestore/remotestore.go:63]: Remove

File: bob/bobfile/bobfile.go [hunk 8]
```diff
@@ -162,6 +171,31 @@ func bobfileRead(dir string) (_ *Bobfile, err error) {
...
+		case project.Local:
+			// Do nothing
+		case project.Remote:
+			// Initialize remote store
+			url, err := projectname.Remote()
+			if err != nil {
+				return nil, err
+			}
+
+			println("Using remote store:", url.String())
```
[the remark this answers is not in the document]
Comment (rdnt) [reply bob/bobfile/bobfile.go]: Changed to boblog.V with level 1

File: bob/bobfile/bobfile.go (hunk 2 above)
Comment (rdnt) [reply bob/bobfile/bobfile.go]: Added.

File: pkg/store/sync.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+func Sync(ctx context.Context, src, dst Store, id string) (err error) {
+	defer errz.Recover(&err)
+
+	found, err := exists(ctx, src, id)
+	if !found {
+		return ErrArtifactNotFoundinSrc
+	}
+	errz.Fatal(err)
+
+	// found, err = exists(ctx, dst, id)
```
[the remark this answers is not in the document]
Comment (rdnt) [reply pkg/store/sync.go:21]: Are we sure we want to introduce latency here? The remote artifact store will not process an upload request if the artifact already exists upstream.

[the remark this answers is not in the document]
Comment (rdnt) [reply pkg/store/sync.go]: Added check.

**Pull request merged**